File: c/test_bcast.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <math.h>
#include "crestacoll.h"

int main(int argc, char *argv[])
{
   int *sendbuf, *testbuf;
   double t0, t1, time, tave, tmax, tmin;
   int ntask, my_id, root;
   int n, imsg, i, iter, maxiter, large_msg=8192;
   char filename[30];
   FILE *fp;

   MPI_Init(&argc, &argv);
   MPI_Comm_rank(MPI_COMM_WORLD, &my_id);
   MPI_Comm_size(MPI_COMM_WORLD, &ntask);
   root = ntask-1;
  
   if (my_id==0) {
      sprintf(filename, "cresta_bcast%i.dat", ntask);
      fp=fopen(filename,"a");
      fprintf(fp, "# Msg [B]    t_max [us]  t_min [us]  t_ave [us] \n");
   }
   
/* benchmark loop over message sizes */

   for(imsg=0; imsg<=16; imsg+=2){ 
     n = pow(2, imsg); 
     sendbuf = (int*) malloc(n*sizeof(int));
     testbuf = (int*) malloc(n*sizeof(int)); 
     if (n <= large_msg) {
       maxiter = 50;
     } else {
       maxiter = 10;
     }
     if (my_id==root) {
       for (i=0; i<n; i++) {
          sendbuf[i] = my_id*10 + i;
          testbuf[i] = sendbuf[i];
       }
     } else {
       for (i=0; i<n; i++) {
          sendbuf[i] = 0;
          testbuf[i] = 0;
       }
     }
     MPI_Barrier(MPI_COMM_WORLD);
     /* time the collective, repeated by maxiter times */
     time = 0.0;
     for (iter=1; iter<=maxiter; iter++) {
        t0 = MPI_Wtime();
        CRESTA_Bcast(sendbuf, n, MPI_INT, root, MPI_COMM_WORLD); 
        t1 = MPI_Wtime() - t0;
        time += t1;
     } 
     time = time / (double) maxiter;
     /* verify correctness by comparing with MPI */
     MPI_Bcast(testbuf, n, MPI_INT, root, MPI_COMM_WORLD);
     MPI_Barrier(MPI_COMM_WORLD);
     for (i=0; i<n; i++) {
        if (sendbuf[i] != testbuf[i]) {
           printf("Error in CRESTA_Bcast, n=%i, my_id=%i, %i %i \n",
                   sizeof(n)*n, my_id, sendbuf[i], testbuf[i]);
           MPI_Barrier(MPI_COMM_WORLD);
           MPI_Abort(MPI_COMM_WORLD, 1);
        }
     }   
     /* gather and print timings */
     MPI_Reduce(&time, &tmax, 1, MPI_DOUBLE, MPI_MAX, 0, 
                MPI_COMM_WORLD);
     MPI_Reduce(&time, &tmin, 1, MPI_DOUBLE, MPI_MIN, 0,
                MPI_COMM_WORLD);
     MPI_Reduce(&time, &tave, 1, MPI_DOUBLE, MPI_SUM, 0,
                MPI_COMM_WORLD); 
     if (my_id==0) {
        fprintf(fp, "%i %15.3f %15.3f %15.3f \n", 
                sizeof(n)*n, 1.0e6*tmax, 1.0e6*tmin, 
                1.0e6*tave/((double) ntask));
     } 
     free(sendbuf);
     free(testbuf);
     sendbuf=NULL;
     testbuf=NULL;
   }  /* end benchmark loop */
      
   if (my_id==0) fclose(fp); 
   MPI_Finalize(); 
   return 0; 
}

File: ftn/test_bcast.f90
program collectives_benchmark
  use mpi
!  use os_collectives, only : cresta_win_init, cresta_win_finalize
  use crestacoll
  implicit none
  integer, dimension(:), allocatable :: cresta_send, cresta_recv, &
       test_send, test_recv
  real(8) :: time, t0, t1, tmax, tmin, tave
  integer :: n, imsg, i, cnt, iter, maxiter
  integer, parameter :: large_msg = 8192
  character(len=25) :: filename 
  integer :: my_id, ntasks, rc, tmaxloc, tminloc, root


  call mpi_init(rc)
  call mpi_comm_rank(mpi_comm_world, my_id, rc)
  call mpi_comm_size(mpi_comm_world, ntasks, rc)
  root = ntasks-1

! this line is needed when using the OS communication with static windows
!  call cresta_win_init(mpi_comm_world, rc)

  write(filename,'(A12,I0,A4)') 'cresta_bcast', ntasks, '.dat'
  if(my_id==0) open(12,file=filename,form='formatted')

  do imsg = 0,24,4
     n = 2**imsg
     allocate (cresta_send(n), cresta_recv(n))
     allocate (test_send(n), test_recv(n))

     if (n <= large_msg) then
         maxiter = 100
     else
         maxiter = 10	
     end if
     if (my_id == root) then
        do i = 1, n
            cresta_send(i) = my_id*10 + i
        end do
     end if
     test_send = cresta_send
     test_recv = 0
     cresta_recv = 0
     call mpi_barrier(mpi_comm_world, rc)
     time = 0.0d0
     do iter = 0, maxiter
       t0 = gettime()      
       call CRESTA_Bcast(cresta_send, n, mpi_integer, &
                            root, mpi_comm_world, rc)
       t1 = gettime()
       if (iter > 0) time = time + (t1 - t0)
     end do
     time = time / real(maxiter,8)
! compare against MPI
     call MPI_Bcast(test_send, n, mpi_integer, &
                            root, mpi_comm_world, rc)
     if (any(cresta_send(:) /= test_send(:))) then
        write(*,*) ' Error in CRESTA_Bcast'
        do i = 1, n
           if (cresta_send(i) /= test_send(i)) then
              write(*,'(4I10)') my_id, i, cresta_send(i), test_send(i)
           end if
        end do
        call mpi_barrier(mpi_comm_world, rc)
        call mpi_abort(mpi_comm_world, 1, rc)
     end if

     call mpi_reduce(time, tmax, 1, mpi_double_precision, mpi_max, 0, mpi_comm_world, rc)
     call mpi_reduce(time, tmin, 1, mpi_double_precision, mpi_min, 0, mpi_comm_world, rc)
     call mpi_reduce(time, tave, 1, mpi_double_precision, mpi_sum, 0, mpi_comm_world, rc)

     if(my_id==0)  then
        write(12,'(I12,G15.6,G15.6,G15.6)') &
             n, tmax*1.0d6, tmin*1.0d6, 1.0d6*tave/real(ntasks,8)
     end if
     deallocate(cresta_send, cresta_recv)
     deallocate(test_send, test_recv)
  end do
  if(my_id==0) close(12)

!  call cresta_win_finalize(rc)
  call mpi_finalize(rc)

end program collectives_benchmark
